gpustack/worker/serve_manager.py

```diff
@@ -55,62 +54,305 @@
                 await asyncio.sleep(10)
             except Exception as e:
-                logger.error(f"Error while monitoring instances: {e}")
+                logger.error(f"Error restarting model instances: {e}")
                 await asyncio.sleep(5)
 
+    def sync_model_instances_state(self):  # noqa: C901
+        """
+        Synchronize model instances' state.
+
+        - If the provision process is still alive, skip.
+        - If the workload is still launching, skip.
+        - If the workload is not existed, unhealthy, inactive or failed, update the model instance state to ERROR.
+        - If everything is fine, update the model instance state to RUNNING.
+        """
+
+        model_instances = self._clientset.model_instances.list(
+            params={
+                "work_id": self._worker_id,
+            }
+        )
+        for model_instance in model_instances.items:
+            # Skip if the provision process has not exited yet.
+            if self._is_provisioning(model_instance):
+                logger.trace(
+                    f"Model instance {model_instance.name} is provisioning. Skipping sync."
+                )
+                continue
+
+            # Skip if the workload is still launching.
+            workload = get_workload(model_instance.name)
+            if workload and workload.state in [
+                WorkloadStatusStateEnum.PENDING,
+                WorkloadStatusStateEnum.INITIALIZING,
+            ]:
+                logger.trace(
+                    f"Model instance {model_instance.name} workload is still launching. Skipping sync."
+                )
+                continue
+
+            is_main_worker = model_instance.worker_id == self._worker_id
+
+            # Update model instance state to ERROR if the workload is not existed, unhealthy, inactive or failed.
+            if not workload or workload.state in [
+                WorkloadStatusStateEnum.UNHEALTHY,
+                WorkloadStatusStateEnum.INACTIVE,
+                WorkloadStatusStateEnum.FAILED,
+            ]:
+                # Only if not in ERROR state yet.
+                if model_instance.state != ModelInstanceStateEnum.ERROR:
+                    with contextlib.suppress(NotFoundException):
+                        # Get patch dict for main worker.
+                        if is_main_worker:
+                            patch_dict = {
+                                "state": ModelInstanceStateEnum.ERROR,
+                                "state_message": "Inference server exited or unhealthy.",
+                            }
+                        # Get patch dict for subordinate worker.
+                        else:
+                            sw_pos = next(
+                                (
+                                    i
+                                    for i, sw in enumerate(
+                                        model_instance.distributed_servers.subordinate_workers
```

Comment:
Do we need null checks for this

Reply:
@gitlawr  https://github.com/gpustack/gpustack/pull/3052/files#r2472264894